apps/workshop/projects/CroSwap/modules/CroSwap.ts

```diff
@@ -0,0 +1,152 @@
+  name: 'CroSwap',
+  chain: 'cronos',
+  type: 'staking',
+
+  /**
+   * Fetches the addresses of all involved tokens (supplied, rewards, borrowed, etc)
+   *
+   * @param context
+   * @returns Address[]
+   */
+  async preloadTokens(ctx) {
+    return await preloadTokensFromUrl(ctx, CROSWAP_SUBGRAPH_URL);
+  },
+
+  /**
+   * Returns full pool list
+   *
+   * @param context
+   * @returns Pool[]
+   */
+  async fetchPools(ctx: FetchPoolsContext) {
+    return await fetchPoolsFromUrl(ctx, ctx.tokens, CROSWAP_FARMS_URL);
+  },
+
+  /**
+   * Returns user positions for all pools
+   *
+   * @param ctx Context
+   * @returns UserPosition[]
+   */
+  async fetchUserPositions(ctx) {
+    const { pools, user, ethcall, ethcallProvider, ethers } = ctx;
+    const farms = new ethcall.Contract(CROSWAP_FARMS, farmsAbi);
+
+    const positions = pools.flatMap(async (pool) => {
+      const token0 = pool.supplied?.[0].token;
+      if (!token0) return void 0;
+
+      const token1 = pool.supplied?.[1].token;
+      if (!token1) return void 0;
+
+      const lpTokenContract = new ethcall.Contract(pool.id, pairAbi);
+
+      const [poolId, balance, reserves, totalSupply] = (await ethcallProvider.all([
+        farms.getPoolIdforLP(pool.id),
+        lpTokenContract.balanceOf(user),
+        lpTokenContract.getReserves(),
+        lpTokenContract.totalSupply(),
+      ])) as [BigNumber, BigNumber, Reserves, BigNumber];
+
+      const [userInfo, pending] = (await ethcallProvider.all([
+        farms.getUserInfo(poolId, user),
+        farms.pendingRewards(poolId, user),
+      ])) as [UserInfo, BigNumber];
+
+      // total balance of lp tokens in wallet & staked
+      const totalBalance = balance.add(userInfo.amount);
+
+      // dont continue if no balance
+      if (totalBalance.eq(0)) return void 0;
+
+      const reserve0 = FixedNumber.fromValue(reserves._reserve0, 18);
+      const reserve1 = FixedNumber.fromValue(reserves._reserve1, 18);
+
+      // get ratio of total supply
+      const ratioOfTotalSupply = FixedNumber.fromValue(totalBalance).divUnsafe(
+        FixedNumber.from(totalSupply),
+      );
+
+      const token0Balance = parseFloat(
+        formatUnits(reserve0.mulUnsafe(ratioOfTotalSupply).toHexString(), token0.decimals),
+      );
+      const token1Balance = parseFloat(
+        formatUnits(reserve1.mulUnsafe(ratioOfTotalSupply).toHexString(), token1.decimals),
+      );
+
+      return <UserPosition>{
+        id: pool.id,
+        supplied: [
+          {
+            token: pool.supplied?.[0].token,
+            //sucks this is a number
+            balance: token0Balance,
+          },
+          {
+            token: pool.supplied?.[1].token,
+            //sucks this is a number
+            balance: token1Balance,
+          },
+        ],
+        rewarded: [
+          {
+            token: pool.rewarded?.[0].token,
+            //sucks this is a number
+            balance: parseFloat(formatUnits(pending, 18)),
```

Comment:
due to how the dependencies work with our build system, could you change these so that its using the injected `ethers` from ctx, instance instead of importing directly?

```diff
-import { formatUnits } from 'ethers/lib/utils';

  ...
  ...
-  balance: parseFloat(formatUnits(pending, 18)),
+  balance: parseFloat(ethers.utils.formatUnits(pending, 18)),
```